src/pubsub.c

```diff
@@ -285,18 +317,25 @@ int pubsubSubscribeChannel(client *c, robj *channel, pubsubtype type) {
 /* Unsubscribe a client from a channel. Returns 1 if the operation succeeded, or
  * 0 if the client was not subscribed to the specified channel. */
 int pubsubUnsubscribeChannel(client *c, robj *channel, int notify, pubsubtype type) {
+    dict *d;
     dictEntry *de;
     list *clients;
     listNode *ln;
     int retval = 0;
+    unsigned int slot = 0;
 
     /* Remove the channel from the client -> channels hash table */
     incrRefCount(channel); /* channel may be just a pointer to the same object
                             we have in the hash tables. Protect it... */
     if (dictDelete(type.clientPubSubChannels(c),channel) == DICT_OK) {
         retval = 1;
         /* Remove the client from the channel -> clients list hash table */
-        de = dictFind(*type.serverPubSubChannels, channel);
+        if (server.cluster_enabled && type.shard) {
+            slot = c->slot;
```

Comment:
here we should recalculate the slot if `c->slot` is `-1` (`SUNSUBSCRIBE` without specific channel name means sub all).

Reply:
I would be good to have a test case covering this...